src/wrapper_template.h

```diff
@@ -177,19 +187,200 @@ inline nb::object wrap_multiargument_function(const MultiargumentFunc& func, con
-      nb::isinstance<nb::ndarray<int32_t>>(array) || nb::isinstance<nb::ndarray<uint32_t>>(array) ||
-      nb::isinstance<nb::ndarray<int64_t>>(array) || nb::isinstance<nb::ndarray<uint64_t>>(array)) {
-    return true;
+  for (const auto& argument : input) {
+    if (nb::isinstance<nb::list>(argument)) {
+      auto list = nb::cast<nb::list>(argument);
+      shape.push_back(nb::len(list));
+      std::vector<nb::object> tmp;
+      tmp.reserve(nb::len(list));
+      for (auto item : list) tmp.push_back(nb::cast(item));
+      expanded_inputs.push_back(std::move(tmp));
+    } else if (nb::isinstance<nb::ndarray<>>(argument)) {
+      if (check_int_dtype(argument)) {
+        auto array = nb::cast<nb::ndarray<long long, nb::shape<-1>>>(argument);
+
+        shape.push_back(array.size());
+        std::vector<nb::object> tmp;
+        tmp.reserve(array.size());
+        for (size_t i = 0; i < array.size(); ++i) {
+          tmp.push_back(nb::cast(array(i)));
+        }
+        expanded_inputs.push_back(std::move(tmp));
+      } else {
+        auto array = nb::cast<nb::ndarray<double, nb::shape<-1>>>(argument);
+        shape.push_back(array.size());
+        std::vector<nb::object> tmp;
+        tmp.reserve(array.size());
+        for (size_t i = 0; i < array.size(); ++i) {
+          tmp.push_back(nb::cast(array(i)));
+        }
+        expanded_inputs.push_back(std::move(tmp));
+      }
+    } else if (nb::isinstance<nb::float_>(argument) || nb::isinstance<nb::int_>(argument)) {
+      shape.push_back(1);
+      expanded_inputs.push_back({argument});
+    } else {
+      // Handle unsupported types
+      throw nb::type_error("Input must be a float, int, list, or 0-D/1-D NumPy array.");
+    }
+    if (shape.back() <= 0) {
+      return nb::ndarray<double, nb::numpy>(nullptr, {0}).cast();  // return empty array
+    }
   }
-  return false;
+
+  std::function<nb::object(size_t, std::vector<nb::object>&)> build;
+  build = [&](size_t depth, std::vector<nb::object>& current) -> nb::object {
+    if (depth == expanded_inputs.size()) {
+      std::vector<std::variant<double, int>> args;
+      args.reserve(current.size());
+      for (const auto& obj : current) {
+        if (nb::isinstance<nb::float_>(obj)) {
+          args.emplace_back(nb::cast<double>(obj));
+        } else if (nb::isinstance<nb::int_>(obj)) {
+          args.emplace_back(nb::cast<int>(obj));
+        } else {
+          throw nb::type_error("All arguments must be int or float at the deepest level.");
+        }
+      }
+      return nb::cast(func(args));
+    }
+    nb::list out;
+    for (auto& val : expanded_inputs[depth]) {
+      current.push_back(val);
+      out.append(build(depth + 1, current));
+      current.pop_back();
+    }
+    return out;
+  };
+  std::vector<nb::object> current;
+  return build(0, current);
+}
+
+inline nb::object wrap_cartesian_product_function_new(const MultiargumentFunc& func,
```

Comment:
that function has grown too much, can you split it down into smaller, more manageable pieces ?